ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/MediaPlayerService.kt

```diff
@@ -136,6 +148,99 @@ class MediaPlayerService : Service() {
+        rootHints: Bundle?
+    ): MediaBrowserServiceCompat.BrowserRoot {
+
+        // Returns a root ID that clients can use with onLoadChildren() to retrieve
+        // the content hierarchy. Note that this root isn't actually displayed.
+        return MediaBrowserServiceCompat.BrowserRoot(MEDIA_BROWSER_ROOT_ID, null)
+    }
+
+    override fun onLoadChildren(
+        parentMediaId: String,
+        result: MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>>
+    ) {
+
+        val mediaItems: MutableList<MediaBrowserCompat.MediaItem> = mutableListOf()
+
+        if (MEDIA_BROWSER_ROOT_ID == parentMediaId) {
+            // Build the MediaItem objects for the top level,
+            // and put them in the mediaItems list...
+
+            var albumList: MediaDescriptionCompat.Builder = MediaDescriptionCompat.Builder()
+            albumList.setTitle("Browse Albums").setMediaId(MEDIA_BROWSER_ALBUM_LIST_ROOT)
+            mediaItems.add(
+                MediaBrowserCompat.MediaItem(
+                    albumList.build(),
+                    MediaBrowserCompat.MediaItem.FLAG_BROWSABLE
+                )
+            )
+        } else if (MEDIA_BROWSER_ALBUM_LIST_ROOT == parentMediaId) {
+            executorService.execute {
+                val musicService = getMusicService()
+
+                val musicDirectory: MusicDirectory = musicService.getAlbumList2(
+                    "alphabeticalByName", 10, 0, null
+                )
+
+                for (item in musicDirectory.getAllChild()) {
+                    var entryBuilder: MediaDescriptionCompat.Builder =
+                        MediaDescriptionCompat.Builder()
+                    entryBuilder
+                        .setTitle(item.title)
+                        .setMediaId(MEDIA_BROWSER_ALBUM_PREFIX + item.id)
+                    mediaItems.add(
+                        MediaBrowserCompat.MediaItem(
+                            entryBuilder.build(),
+                            MediaBrowserCompat.MediaItem.FLAG_BROWSABLE
+                        )
+                    )
+                }
+                result.sendResult(mediaItems)
+            }
+            result.detach()
+            return
+        } else if (parentMediaId.startsWith(MEDIA_BROWSER_ALBUM_PREFIX)) {
+            executorService.execute {
+                val musicService = getMusicService()
+                val id = parentMediaId.substring(MEDIA_BROWSER_ALBUM_PREFIX.length)
+
+                val albumDirectory = musicService.getAlbum(
+                    id, "", false
+                )
+                for (item in albumDirectory.getAllChild()) {
+                    var extras = Bundle()
+
+                    var baos = ByteArrayOutputStream()
+                    var oos = ObjectOutputStream(baos)
+                    oos.writeObject(item)
+                    oos.close()
+                    extras.putByteArray(MEDIA_BROWSER_EXTRA_ENTRY_BYTES, baos.toByteArray())
```

Comment:
This is where we add the song list to display after someone selects an album. Note that `item` implements `Serializable` so I should just be able to call `extras.putSerializable(...)` but when I tried that it crashes. Not sure if we don't implement something in `Serializable` correctly. I cheat and use a byte array, I'd need to clean this up.